File: ios/App/App/FocusedWidgetSyncPlugin.swift
import Foundation
import Capacitor
import WidgetKit

// Ponte tra la sessione Supabase nella WebView (src/lib/nativeWidgetSync.js) e
// l'App Group condiviso che il target Widget Extension (Fase 3) legge per
// chiamare /api/widget-data. L'App Group va creato in Xcode (Signing &
// Capabilities su entrambi i target, app + widget extension) quando si
// aggiunge il target — fino ad allora syncSession fallisce in modo pulito.
@objc(FocusedWidgetSyncPlugin)
public class FocusedWidgetSyncPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "FocusedWidgetSyncPlugin"
    public let jsName = "FocusedWidgetSync"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "syncSession", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearSession", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "refreshWidgets", returnType: CAPPluginReturnPromise),
    ]

    static let appGroupId = "group.org.focusedapp.app"

    @objc func syncSession(_ call: CAPPluginCall) {
        guard let defaults = UserDefaults(suiteName: Self.appGroupId) else {
            call.reject("App Group non ancora configurato in Xcode")
            return
        }
        defaults.set(call.getString("accessToken"), forKey: "access_token")
        defaults.set(call.getString("refreshToken"), forKey: "refresh_token")
        defaults.set(call.getString("expiresAt"), forKey: "expires_at")
        call.resolve()
    }

    @objc func clearSession(_ call: CAPPluginCall) {
        guard let defaults = UserDefaults(suiteName: Self.appGroupId) else {
            call.reject("App Group non ancora configurato in Xcode")
            return
        }
        defaults.removeObject(forKey: "access_token")
        defaults.removeObject(forKey: "refresh_token")
        defaults.removeObject(forKey: "expires_at")
        call.resolve()
    }

    @objc func refreshWidgets(_ call: CAPPluginCall) {
        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }
        call.resolve()
    }
}

File: android/app/src/main/java/org/focusedapp/app/FocusedWidgetSyncPlugin.kt
package org.focusedapp.app

import android.content.Context
import android.content.Intent
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

// Ponte tra la sessione Supabase nella WebView (src/lib/nativeWidgetSync.js) e
// lo storage condiviso che i widget Glance (Fase 4) leggono per chiamare
// /api/widget-data. Stesso processo/app, quindi SharedPreferences basta —
// niente App Group come su iOS.
@CapacitorPlugin(name = "FocusedWidgetSync")
class FocusedWidgetSyncPlugin : Plugin() {
    companion object {
        const val PREFS_NAME = "focused_widget_prefs"
        const val REFRESH_ACTION = "org.focusedapp.app.REFRESH_WIDGETS"
    }

    @PluginMethod
    fun syncSession(call: PluginCall) {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        prefs.edit()
            .putString("access_token", call.getString("accessToken"))
            .putString("refresh_token", call.getString("refreshToken"))
            .putString("expires_at", call.getString("expiresAt"))
            .apply()
        call.resolve()
    }

    @PluginMethod
    fun clearSession(call: PluginCall) {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        prefs.edit().clear().apply()
        call.resolve()
    }

    @PluginMethod
    fun refreshWidgets(call: PluginCall) {
        // I singoli AppWidgetProvider Glance (Fase 4) ascoltano questo broadcast
        // per aggiornarsi subito invece di aspettare il refresh periodico di WorkManager.
        context.sendBroadcast(Intent(REFRESH_ACTION).setPackage(context.packageName))
        call.resolve()
    }
}
